### PR TITLE
Add github action to be able to release on published tag

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
This permit to release on pypi if a value for ``PYPI_API_TOKEN`` is set in github settings. The release is done when a release is created/published in github interface.

Relates to #575

### CODE REVIEW

**Comment by sigmavirus24** [comment]
This should be a protected workflow so only trusted folks can do this, no?

**Comment by Pierre-Sassoulas** [comment]
> This should be a protected workflow so only trusted folks can do this, no?

The workflow is triggered by [creating a release](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) (in github interface from this url https://github.com/PyCQA/pydocstyle/releases) so only members that can release a package can trigger this workflow.

**Comment by sigmavirus24** [comment]
I wasn't sufficiently clear. There are often cases where you don't want everyone with permission to create a release to be able to publish to PyPI. No one can pip install from GitHub and get malicious code (although some people apparently do install artifacts from there which boggles my mind). Publishing to PyPI should be more severely restricted. I thought @sethmlarson had written a blog post about this but I'm not seeing it.

**Comment by Pierre-Sassoulas** [comment]
I don't see the purpose of a release without a pypi publication alongside it, so for me those permissions were equivalent. Let me know if I need to modify something :)

> some people apparently do install artifacts from there which boggles my mind

I would have made the mistake 15 years ago. It's probably autodidacts with no formal training who were never introduced to package management. When you come from a Windows background you're used to searching for an executable to install after searching for the name of the software online, and github releases are what's closest to this.

**Comment by sethmlarson** [comment]
@sigmavirus24 Article you're looking for is here: https://sethmlarson.dev/blog/security-for-package-maintainers

**Comment by sigmavirus24** [comment]
> for me those permissions were equivalent

They shouldn't always be equivalent. That's my point. Also, GitHub releases are trash for so many reasons and create extra burden on maintainers. They shouldn't be the vehicle to publishing to PyPI

**Comment by Pierre-Sassoulas** [comment]
It's also possible to launch a github actions on tag creation directly but there would still be the problem of who has right to create tag. If you don't want to mix pypi / github rights then the solution is probably a local script launched by someone with pypi's release right.

**Comment by samj1912** [comment]
@Pierre-Sassoulas - I have created two environments `pypi-dev` (which has tokens for test.pypi) and `pypi-prod` (which has tokens for pypi) both of which are gated and require approvers from the current admins (me and @Nurdok) who have release permissions on PyPi. Could you follow something similar to https://timheuer.com/blog/add-approval-workflow-to-github-actions/ to add these environments to the workflow? 

I have created project specific tokens and added them as env vars to these environments.

**Comment by samj1912** [comment]
Separately - it would have been great if we could have versions also driven by the Github tags instead having to manually bump them similar to https://github.com/cruft/cruft/blob/0c6de85d974197969c0e65913857eaa36b788e5e/.github/workflows/pypi_publish.yml#L22

but any improvement to the release process is great. Thanks for this PR! 🙌

**Comment by Pierre-Sassoulas** [comment]
@samj1912 I don't know if I understood your first link correctly, Would [342cc17](https://github.com/PyCQA/pydocstyle/pull/582/commits/342cc1780f60ca030d39d906b7bfc47f396690a6) limit the person that can launch the job ? In order to test we need to create a tag and publish a release which is... something that I'd try on another smaller repo, personally 😄 

Second link look nice too but I have very limited available time and I'm not knowledgeable enough about poetry right now.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        with:
+          python-version: ${{ env.DEFAULT_PYTHON }}
+          check-latest: true
+      - name: Install requirements
+        run: |
+          python -m pip install -U pip twine wheel
+          python -m pip install -U "setuptools>=56.0.0"
+      - name: Build distributions
+        run: |
+          python setup.py sdist bdist_wheel
```
Comment (adamjstewart) [inline .github/workflows/release.yml:39]: Direct invocation of `setup.py` is [deprecated](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html), current [best practice](https://packaging.python.org/en/latest/tutorials/packaging-projects/#generating-distribution-archives) is to use `python -m build` instead.

**Review by adamjstewart — COMMENTED** [review]
Most of the projects I maintain use [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) and [actions/upload-artifact](https://github.com/actions/upload-artifact) instead of manual `build` and `twine`. May be more reliable since they are maintained by PyPA or GitHub themselves.

**Comment by Pierre-Sassoulas** [comment]
> Most of the projects I maintain use [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) and [actions/upload-artifact](https://github.com/actions/upload-artifact) instead of manual build and twine. May be more reliable since they are maintained by PyPA or GitHub themselves.

This looks pretty cool, are wheel properly created for al operating systems with ``python -m build`` ? Because otherwise we should definitely use what you suggested. (I'm using this release workflow in a number of places so I appreciate the review, thank you @adamjstewart  :) )

**Comment by adamjstewart** [comment]
> are wheel properly created for al operating systems with `python -m build`?

In the last couple of years, PyPA completely overhauled their support for non-default build backends. Build "backends" include things like distutils, setuptools, flit, hatchling, poetry, etc. Build "frontends" include things like build and pip. One of the advantages of build and pip is that they support all of the above build backends, whereas setup.py only supports distutils/setuptools. Also setup.py is deprecated as I mentioned.

`python -m build` basically does the same thing that `python setup.py build` used to do. For pure-Python projects like pydocstyle, it creates a single universal wheel for Python 3. For Python libraries that depend on external libraries or that require compilation, it builds a Python-version and platform-specific wheel, and you need to run this command on all versions/platforms (e.g., via matrix in GitHub Actions) and upload each one separately. 

cibuildwheel is probably overkill for this project since we don't need a separate wheel for each Python version/platform. Let's stick with build/twine instead.

**Comment by Pierre-Sassoulas** [comment]
Thanks a lot for the explanation, I only have pure python project right now but I'll keep that in mind 👍

**Review by samj1912 — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by samj1912** [comment]
Thank you so much for this!

**Pull request merged**